### PR TITLE
RFC: Deduplicate tag and attribute names

### PR DESCRIPTION
**Pull request by adamreichold**
This is a follow-up to #80 which reboots deduplication of tag and attribute names using the same technique with a separate sorted list of indexes as employed there.

Since `ExpandedNameIndexed` does contain any allocations itself, the deduplicate has also become easier as we can produce such a value for comparison with a most likely unnecessary allocation. Furthermore, since `ExpandedNameIndexed` would refer to deduplicated namespaces since #80, we still have a high chance of actually deduplicating names across the document.

For `greater_london.osm`, this reduces memory consumption from 6.7 GB to 5.5 GB, i.e. by over 16%. (`doc.expanded_names.values` has 17 entries in this case.)

For the CSW response mentioned in #80, this further reduces memory consumption 108 MB to 100 MB, i.e. by almost 8%.

Benchmarks are still a mixed when compared to `master`, i.e.

```
 name                                    master ns/iter  dedup-tag-attribute-names ns/iter  diff ns/iter   diff %  speedup 
 large_roxmltree                         4,020,847       3,856,116                              -164,731   -4.10%   x 1.04 
 medium_roxmltree                        804,519         1,039,112                               234,593   29.16%   x 0.77 
 roxmltree_iter_children                 3,586           3,027                                      -559  -15.59%   x 1.18 
 roxmltree_iter_descendants_expensive    128,694         128,533                                    -161   -0.13%   x 1.00 
 roxmltree_iter_descendants_inexpensive  44,570          43,330                                   -1,240   -2.78%   x 1.03 
 tiny_roxmltree                          4,860           5,819                                       959   19.73%   x 0.84 
```

while parsing has regressed as expected while iteration has improved when comparing this directly with #80, i.e.

```
 name                                    dedup-namespaces ns/iter  dedup-tag-attribute-names ns/iter  diff ns/iter   diff %  speedup 
 large_roxmltree                         3,431,061                 3,856,116                               425,055   12.39%   x 0.89 
 medium_roxmltree                        810,177                   1,039,112                               228,935   28.26%   x 0.78 
 roxmltree_iter_children                 3,574                     3,027                                      -547  -15.30%   x 1.18 
 roxmltree_iter_descendants_expensive    128,552                   128,533                                     -19   -0.01%   x 1.00 
 roxmltree_iter_descendants_inexpensive  44,731                    43,330                                   -1,401   -3.13%   x 1.03 
 tiny_roxmltree                          4,918                     5,819                                       901   18.32%   x 0.85
```

So in summary, this yields much larger memory savings that apply to more documents compared to #80, but it is also the stronger trade-off against parsing speed.

Considering that even 'greater_londom.osm` uses only 17 unique names, we might actually be able to optimize this by adaptive using a linear search instead of the binary search which should be friendlier towards branch predictors and SIMD optimizations.

### CODE REVIEW

**Comment by RazrFalcon** [comment]
@adamreichold Hi! There were so many changes to this crate recently, that I've completely lost track of them. Would you mind explaining the current algorithm to me? Or maybe even as a comment in the source code itself.

To my understanding, we're using index references for element/attribute names now. And those names are unique as well now? You also did something to namespaces, right? But they were unique before as well.
Would it be useful to use `Rc` instead of indexes? Or lifetimes would be to hard?

**Comment by adamreichold** [comment]
> @adamreichold Hi! There were so many changes to this crate recently, that I've completely lost track of them. Would you mind explaining the current algorithm to me? Or maybe even as a comment in the source code itself.

It is probably best to ignore this PR until we settled what to do with #80: If #80 is dropped, this PR does not really make sense at all. (It does include the commits from #80 because it depends on them.) If #80 is merged, this could be going further in the same direction, but it is not an obvious thing to do.
 
> To my understanding, we're using index references for element/attribute names now. And those names are unique as well now? You also did something to namespaces, right? But they were unique before as well. Would it be useful to use `Rc` instead of indexes? Or lifetimes would be to hard?

Using `Rc` (as I did before) has significant downsides as it would make `Document: !Send` and add significant overhead due to reference count updates (which are unnecessary as everything is tied to the lifetime of `Document` in any case). Therefore, I think using indexes is really the best choice, especially as it also gives us the choice of trading of index size (`u16`, `u32`, `u64`) versus maximum number of entities.

File: src/lib.rs [hunk 1]
```diff
@@ -549,48 +554,118 @@ impl<'input> Namespace<'input> {
...
+            Err(sorted_idx) => {
+                if self.values.len() > core::u16::MAX as usize {
+                    return Err(Error::NamespacesLimitReached);
+                }
+                let idx = self.values.len() as u16;
+                self.values.push(Namespace {
+                    name,
+                    uri: uri.to_cow(),
+                });
+                self.sorted.insert(sorted_idx, idx);
```
[the remark this answers is not in the document]
Comment (adamreichold) [reply src/lib.rs:598]: I would say both are valid strategies: Using two vectors provides better data locality and code density but makes insertion asymptotically slower as you mentioned. `BTreeSet` will likely be faster for larger numbers of unique names, but will also need more code and have worse data locality.

So I'd say we need to measure and to do that determine the target, i.e. how many unique names and how many usages of names.

[the remark this answers is not in the document]
Comment (adamreichold) [reply src/lib.rs:598]: (Note that this is actually part of #80.)

[the remark this answers is not in the document]
Comment (adamreichold) [reply src/lib.rs:598]: (Note that this is actually part of #80.)

File: src/lib.rs [hunk 2]
```diff
@@ -549,48 +554,118 @@ impl<'input> Namespace<'input> {
...
     }
 }
 
-struct Namespaces<'input>(Vec<Namespace<'input>>);
+#[derive(Default)]
+struct Namespaces<'input> {
+    // Deduplicated namespace values used throughout the document
+    values: Vec<Namespace<'input>>,
+    // Indices into the above in tree order as the document is parsed
+    tree: Vec<u16>,
```
[the remark this answers is not in the document]
Comment (adamreichold) [reply src/lib.rs:570]: How do you feel about `parsed`? (Note that this is part of #80.)

File: src/parse.rs [hunk 3]
```diff
@@ -966,11 +989,27 @@ fn process_text<'input>(
     Ok(())
 }
 
-enum BorrowedText<'input, 'temp> {
+pub(crate) enum BorrowedText<'input, 'temp> {
     Input(&'input str),
     Temp(&'temp str),
 }
 
+impl<'input, 'temp> BorrowedText<'input, 'temp> {
```
[the remark this answers is not in the document]
Comment (adamreichold) [reply src/parse.rs:997]: This is part of #80 where it required to delay taking onwership of normalized namespace URI. (The added methods are code blocks which would otherwise be repeated in both `append_text` and `Namespaces::push_ns`.)

File: src/lib.rs [hunk 4]
```diff
@@ -1387,23 +1468,30 @@ impl<'a, 'input: 'a> DoubleEndedIterator for Children<'a, 'input> {
...
     fn new(start: Node<'a, 'input>) -> Self {
+        let from = start.id.get_usize();
+
+        let until = start
+            .d
+            .next_subtree
+            .map(NodeId::get_usize)
+            .unwrap_or(start.doc.nodes.len());
+
+        let nodes = start.doc.nodes[from..until].iter().enumerate();
```
[the remark this answers is not in the document]
Comment (adamreichold) [reply src/lib.rs:1489]: The main problem I had, was that mapping here is not possible with a closure as the type to embed in `Descendants` would become unnameable otherwise.

(As discussed in https://github.com/RazrFalcon/roxmltree/pull/80#discussion_r1028314732, I think these types should eventually go away completely using something like

```rust
pub type Descendants<'a> = impl ExactSizeIterator + DoubleEndedIterator + Clone + fmt::Debug + 'a;

pub fn descendants(&self) -> Descendants {
  ...

  self.doc.nodes[from..until].iter().enumerate().map(...)
}
```

but that will take some time until that feature is stabilized.)

File: src/lib.rs [hunk 5]
```diff
@@ -1387,23 +1468,30 @@ impl<'a, 'input: 'a> DoubleEndedIterator for Children<'a, 'input> {
...
 
 /// Iterator over a node and its descendants.
-#[derive(Clone, Debug)]
+#[derive(Clone)]
 pub struct Descendants<'a, 'input> {
     doc: &'a Document<'input>,
-    current: NodeId,
-    until: NodeId,
+    nodes: core::iter::Enumerate<core::slice::Iter<'a, NodeData<'input>>>,
+    from: usize,
```
[the remark this answers is not in the document]
Comment (adamreichold) [reply src/lib.rs:1475]: Indeed. I will create a separate PR for only this change.

[the remark this answers is not in the document]
Comment (adamreichold) [reply src/lib.rs:1475]: (Note that this is also comes from #80.)

File: src/lib.rs [hunk 6]
```diff
@@ -405,7 +406,8 @@ struct NodeData<'input> {
 
 #[derive(Clone, Debug)]
 struct AttributeData<'input> {
-    name: ExpandedNameIndexed<'input>,
+    /// Indexes into `ExpandedNames::values`
+    name_idx: u32,
```
[the remark this answers is not in the document]
Comment (adamreichold) [reply src/lib.rs:410]: It is never exposed in the crate's API so it would only be us who have to keep this straight, but yes a newtype replace the comment and prevent accidentally indexing into the wrong collection. We should probably start with `namespace_idx` in #80 though.

@RazrFalcon Since it's your project: What are you preferences w.r.t using newtypes to mark the target of the indexes?

[the remark this answers is not in the document]
Comment (RazrFalcon) [reply src/lib.rs:410]: I think it would be a good idea. A bit more verbose, but potentially safer.

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (RazrFalcon) [reply src/lib.rs:598]: @tomjw64 A typical XML has like 1-5 namespaces. I don't think we would hit any performance issues here.

File: src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (RazrFalcon) [reply src/lib.rs:570]: Agree. It threw me off as well. How about `tree_oder` and `sorted_order`?

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (adamreichold) [reply src/lib.rs]: @RazrFalcon If we really want to make this assumption, we can drop the binary search completely as a simple linear search will do and the separate `sorted` list will be unnecessary. It will mean more significant performance degradation for files with a larger number of namespaces though. (For example, that CSW response has 15 namespaces and I regularly process RDF files with a similar number of namespaces.)

**Comment by adamreichold** [comment]
Since #80 was merged still including all of the stuff we are discussing here, I will open a separate PR fixing those things up as to avoid conflating them with the question of whether tag and attribute names should be deduplicated. (Which I have basically started by posting this PR too early. Sorry.)

File: src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (adamreichold) [reply src/lib.rs]: Both are fine with me. @tomjw64 You alright with `tree_order` and `sorted_order`?

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (adamreichold) [reply src/lib.rs]: So, to maybe add some data on this:

`large.plist` sees a small slowdown when the binary search is replaced by a linear search:

```
Benchmark 1: ./stats-linear-search benches/large.plist
  Time (mean ± σ):       5.7 ms ±   0.3 ms    [User: 4.5 ms, System: 0.9 ms]
  Range (min … max):     5.0 ms …   6.7 ms    523 runs
 
Benchmark 2: ./stats-binary-search benches/large.plist
  Time (mean ± σ):       5.5 ms ±   0.4 ms    [User: 4.5 ms, System: 0.9 ms]
  Range (min … max):     4.9 ms …   6.7 ms    535 runs
 
Summary
  './stats-binary-search benches/large.plist' ran
    1.03 ± 0.09 times faster than './stats-linear-search benches/large.plist'
```

which is strange since that file does not use namespaces, so I guess that the simpler linear search leads to worse inlining decisions?

`medium.svg` which does actually have 8 namespaces also sees a slow down:

```
Benchmark 1: ./stats-linear-search benches/medium.svg
  Time (mean ± σ):       2.3 ms ±   0.1 ms    [User: 1.8 ms, System: 0.4 ms]
  Range (min … max):     2.0 ms …   3.1 ms    1190 runs
 
Benchmark 2: ./stats-binary-search benches/medium.svg
  Time (mean ± σ):       2.2 ms ±   0.1 ms    [User: 1.7 ms, System: 0.4 ms]
  Range (min … max):     2.0 ms …   3.0 ms    1430 runs
 
Summary
  './stats-binary-search benches/medium.svg' ran
    1.06 ± 0.09 times faster than './stats-linear-search benches/medium.svg'

```

My CSW response is similarly faster with binary search:

```
Benchmark 1: ./stats-linear-search /home/adam/Downloads/q.xml
  Time (mean ± σ):     484.6 ms ±   3.4 ms    [User: 451.1 ms, System: 33.2 ms]
  Range (min … max):   479.5 ms … 490.8 ms    10 runs
 
Benchmark 2: ./stats-binary-search /home/adam/Downloads/q.xml
  Time (mean ± σ):     477.2 ms ±   8.1 ms    [User: 443.2 ms, System: 33.6 ms]
  Range (min … max):   472.3 ms … 499.1 ms    10 runs
 
Summary
  './stats-binary-search /home/adam/Downloads/q.xml' ran
    1.02 ± 0.02 times faster than './stats-linear-search /home/adam/Downloads/q.xml'
```

So while it may as much due to inlining decisions as due to algorithmic choices, it seems binary search is a bit better than linear search contrary to wait I said above.

Will look into `BTreeSet` next...

[the remark this answers is not in the document]
Comment (RazrFalcon) [reply src/lib.rs]: Feel free to add a new XML sample to the benchmark if needed. Well, as long as it's not copyrighted or something.

**Comment by RazrFalcon** [comment]
Good. Just let me know when a patch will be ready to merge.

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (adamreichold) [reply src/lib.rs]: @tomjw64 When trying to implement it, I noticed that using a `BTreeSet` is not that simple: We need a stable order of the values as we parse the document (otherwise we would need to fix up existing indices as the order changes). Hence I could only replaced `sorted(_order)` with a `BTreeMap<Namespace<'input>, u16>` with stable indexes into the `values` list.

But this would mean that all namespaces which need allocation would have to be allocated twice and this structure would have a larger cache footprint. I am pretty sure this would be worse than shuffling around a few `u16`s on insertion.

[the remark this answers is not in the document]
Comment (adamreichold) [reply src/lib.rs]: So in summary, its seems like a flat list of `u16` and binary search is still the best option?

File: src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (adamreichold) [reply src/lib.rs:410]: This is part of #84

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tomjw64) [reply src/lib.rs]: > @tomjw64 A typical XML has like 1-5 namespaces. I don't think we would hit any performance issues here.

We aren't just using this technique for Namespaces, but also for ExpandedNames in this PR. We are not talking about just 5 namespaces in this case, right? Are we also unconcerned by the potential number of unique expanded names?

[the remark this answers is not in the document]
Comment (tomjw64) [reply src/lib.rs]: > @tomjw64 When trying to implement it, I noticed that using a `BTreeSet` is not that simple: We need a stable order of the values as we parse the document (otherwise we would need to fix up existing indices as the order changes). Hence I could only replaced `sorted(_order)` with a `BTreeMap<Namespace<'input>, u16>` with stable indexes into the `values` list.
> 
> But this would mean that all namespaces which need allocation would have to be allocated twice and this structure would have a larger cache footprint. I am pretty sure this would be worse than shuffling around a few `u16`s on insertion.

I don't dispute that `tree_order` should be a `Vec`, but as for `sorted_order`, could we try  a `BTreeSet<NewType>` where `struct NewType(u32/u16)` and `NewType::cmp` indexes into `values` to make a comparison?

[the remark this answers is not in the document]
Comment (RazrFalcon) [reply src/lib.rs]: @tomjw64 My bad. I thought this code applies only to namespaces.

@adamreichold Yes. Seems to be good enough.

File: src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (adamreichold) [reply src/lib.rs]: Part of #84

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (adamreichold) [reply src/lib.rs]: > Feel free to add a new XML sample to the benchmark if needed. Well, as long as it's not copyrighted or something.

Will do in #84, but I guess I will aim for something in the sub 1MB range as it will going into Git.

> We aren't just using this technique for Namespaces, but also for ExpandedNames in this PR. We are not talking about just 5 namespaces in this case, right? Are we also unconcerned by the potential number of unique expanded names?

> My bad. I thought this code applies only to namespaces.

Well the comment is attached to the place were namespaces are handled, but yes deduplication of expanded names as presented here uses the same technique (but with a `u32` instead of a `u16` index).

I basically agree that for tag and attribute names, the trade-offs are different and much more distinct names are possible. I still think that the flat sorted list is the best approach due to locality, but even with this I repeated the benchmarks against `master` after deduplication of namespaces with the change to `u16` and avoiding redundant allocations was merged and they look rather bad now as the baseline has gotten faster:

```
 name                                    master ns/iter  dedup ns/iter  diff ns/iter   diff %  speedup 
 large_roxmltree                         3,512,675       4,252,045           739,370   21.05%   x 0.83 
 medium_roxmltree                        818,021         1,048,494           230,473   28.17%   x 0.78 
 roxmltree_iter_children                 3,626           3,006                  -620  -17.10%   x 1.21 
 roxmltree_iter_descendants_expensive    121,392         120,818                -574   -0.47%   x 1.00 
 roxmltree_iter_descendants_inexpensive  39,760          37,976               -1,784   -4.49%   x 1.05 
 tiny_roxmltree                          5,211           5,931                   720   13.82%   x 0.88 
```

So I think we should _not_ do this any more.

I am sorry for all the hassle I put you guys though, but the memory savings from 6.7 GB to 5.5 GB are hard to ignore and I hoped this technique would improve things enough to make this workable but alas it does not.

> As for sorted_order, could we try a BTreeSet<NewType> where struct NewType(u32/u16) and NewType::cmp indexes into values to make a comparison?

The problem is that `BTreeSet<NewType>` requires `NewType: Ord`, i.e. implementing `fn cmp(&self, other: &Self) -> Ordering;`, so there is no place to slip in the reference to `values` which is need to resolve `NewType` into something that can be reasonably sorted. (One could try to make `NewType(u16, &[Namespace<'input>])` but that would make it redundantly large and `struct Namespaces` self-referential.)

[the remark this answers is not in the document]
Comment (adamreichold) [reply src/lib.rs]: Also note that this slow down is unlikely to be due to the algorithmic complexity of the flat sorted list as e.g. `large.plist` has 8 unique element and attribute names whereas `fonts.conf` has 9. Only `medium.svg` has 148 names which might explain why it is the worst of the three but even 148 `u16`s are only `(148 * 2 + 63) / 64 = 5` cache lines.

It just seems that deduplicating namespace is a bit of a sweet spot as we already invest quite a bit of effort into their processing so that the extra cost of the deduplication doesn't impact things too badly while their inheritance rules means they will be referenced a lot of (`tree_order` is quite large) even though there are only a few distinct values (`values` is quite small).

[the remark this answers is not in the document]
Comment (tomjw64) [reply src/lib.rs]: Makes sense why the BTreeSet won't work. Super thorough, thanks.
